### PR TITLE
371: typescript refactor

### PR DESCRIPTION
**Pull request by maxb94**

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -52,20 +55,24 @@ const rrfConfig = {
 firebase.initializeApp(firebaseConfig);
 const analytics = firebase.analytics();
 firebase.auth();
+firebase.firestore();
```
Comment (antoineeudes) [inline src/index.js:62]: what is this line doing?

Comment (maxb94) [reply src/index.js:62]: https://react-redux-firebase.com/docs/getting_started.html

It is probably done by `reduxFirestore` behind the scenes but we are no longer using that.

File: src/components/molecules/NavBar/NavBar.tsx [hunk 2]
```diff
@@ -76,7 +74,7 @@ const NavBar: React.FunctionComponent<PropsType> = ({ redirectionUrl }) => {
               />
             </span>
           </Link>
-          {user ? (
+          {profile ? (
```
Comment (antoineeudes) [inline src/components/molecules/NavBar/NavBar.tsx:75]: it should be user here I think. We want to see the avatar with an anonymous user picture when the user doesn't have a profile

File: src/components/molecules/NavBar/NavBar.tsx [hunk 3]
```diff
@@ -112,10 +110,7 @@ const NavBar: React.FunctionComponent<PropsType> = ({ redirectionUrl }) => {
                 onClick={() => setIsProfileModalOpen(true)}
               >
                 <img
-                  src={
-                    users?.[user.uid]?.pictureUrl ||
-                    "/anonymous-profile-icon.jpeg"
-                  }
+                  src={profile.pictureUrl}
```
Comment (antoineeudes) [inline src/components/molecules/NavBar/NavBar.tsx]: profile?.pictureUrl || "/anonymous-profile-icon.jpeg"

File: src/components/organisms/Room/Room.tsx [hunk 4]
```diff
@@ -73,12 +80,10 @@ const Room = ({ roomName, setUserList, capacity = undefined }) => {
     return () => {
       if (localRoom && localRoom.localParticipant.state === "connected") {
         localRoom.localParticipant.tracks.forEach(function (trackPublication) {
-          trackPublication.track.stop();
+          //@ts-ignore
+          trackPublication.track.stop(); //@debt typing does this work?
```
Comment (antoineeudes) [inline src/components/organisms/Room/Room.tsx:85]: what happens here?

File: src/types/Venue.ts [hunk 5]
```diff
@@ -32,4 +32,5 @@ export interface Venue {
   };
   profile_questions: Question[];
   code_of_conduct_questions: Question[];
+  iframeUrl?: string; //@debt typing is this optional
```
Comment (antoineeudes) [inline src/types/Venue.ts:36]: I think it should be optional

File: src/components/venues/Jazzbar/components/ReactionList.tsx [hunk 6]
```diff
@@ -56,12 +56,8 @@ const ReactionList: React.FC<ReactionListProps> = ({
             <div className="partyname-bubble">
               {usersById[message.created_by].partyName}:
             </div>
-            <div
-              className={`message-bubble ${
-                isMessageToTheBand(message) ? "" : "emoji"
-              }`}
-            >
-              {isMessageToTheBand(message)
+            <div className={`message-bubble ${message ? "" : "emoji"}`}>
```
Comment (mathieud) [inline src/components/venues/Jazzbar/components/ReactionList.tsx:59]: Is this right? How can message be undefined here?

File: src/components/molecules/NavBar/NavBar.tsx [hunk 7]
```diff
@@ -1,35 +1,33 @@
...
     venue: state.firestore.data.currentVenue,
     privateChats: state.firestore.ordered.privatechats,
   }));
 
   const now = firebase.firestore.Timestamp.fromDate(new Date());
-  const futureUpcoming = venue?.events?.filter(
-    (e: UpcomingEvent) => e?.ts_utc?.valueOf() > now.valueOf()
-  );
+  const futureUpcoming =
+    venue?.events?.filter((e) => e.ts_utc.valueOf() > now.valueOf()) ?? []; //@debt typing does this exist?
```
Comment (antoineeudes) [inline src/components/molecules/NavBar/NavBar.tsx:30]: no it doesn't

File: src/components/venues/Jazzbar/components/ReactionList.tsx (hunk 6 above)
Comment (mathieud) [reply src/components/venues/Jazzbar/components/ReactionList.tsx:59]: Did you mean  `message.reaction === "messageToTheBand"`

File: src/hooks/useConnectCurrentEvent.ts [hunk 8]
```diff
@@ -22,7 +20,7 @@ const useConnectCurrentEvent = () => {
       collection: "purchases",
       where: [
         ["eventId", "==", eventId],
-        ["userId", "==", user.uid],
+        ["userId", "==", user?.uid],
```
Comment (mathieud) [inline src/hooks/useConnectCurrentEvent.ts:23]: Firebase will throw an error if given `undefined` here

File: src/pages/Account/CodeOfConduct.tsx [hunk 9]
```diff
@@ -37,11 +39,12 @@ const CodeOfConduct: React.FunctionComponent<PropsType> = ({ location }) => {
     mode: "onChange",
   });
   const onSubmit = async (data: CodeOfConductFormData) => {
+    if (!user) return;
     await updateUserProfile(user.uid, data);
     // if we use history.push, the new profile information are not in the redux store
     // when we arrive on `venue/${venueId}/event/${eventId}` and we get redirected again to account/profile
     window.location.assign(
-      `/${venueId && eventId ? `venue/${venueId}/event/${eventId}` : ""}`
+      venueId && eventId ? `venue/${venueId}/event/${eventId}` : ""
```
Comment (mathieud) [inline src/pages/Account/CodeOfConduct.tsx:47]: initial `/` missing?

File: src/types/Firestore.ts [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+interface VenueStatus {
+  currentVenue: boolean;
+  currentEvent: boolean;
+  eventPurchase: boolean;
+  venueChats: boolean;
+  venueEvents: boolean;
+  userPurchaseHistory: boolean;
+}
+
+interface Reaction {
```
Comment (mathieud) [inline src/types/Firestore.ts:21]: This looks wrong. Shouldn't it have the same type as in the creation form?

File: src/types/PartyMapVenue.ts [hunk 11]
```diff
@@ -2,6 +2,10 @@ import { Venue } from "./Venue";
 import { VenueTemplate } from "./VenueTemplate";
 import { RoomData } from "./RoomData";
 
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export const isPartyMapVenue = (val: any): val is PartyMapVenue =>
```
Comment (mathieud) [inline src/types/PartyMapVenue.ts:6]: Can we restrict val to `Venue`?

File: src/components/organisms/Room/Room.tsx (hunk 4 above)
Comment (maxb94) [reply src/components/organisms/Room/Room.tsx:85]: The types for LocalTrack does not have a stop method, so either the types are wrong or this doesn't do anything

File: src/components/molecules/NavBar/NavBar.tsx (hunk 7 above)
Comment (maxb94) [reply src/components/molecules/NavBar/NavBar.tsx:30]: what should it be replaced with?

File: src/components/context/ChatContext.tsx [hunk 12]
```diff
@@ -15,91 +14,74 @@ export const ChatContext = React.createContext<ChatContextType | undefined>(
   undefined
 );
 
-interface GlobalChatMessage {
-  type: "global";
+type Time = firebase.firestore.Timestamp;
+
+type RestrictedMessageType = "room" | "table";
```
[the remark this answers is not in the document]
Comment (maxb94) [reply src/components/context/ChatContext.tsx:19]: Enum's make typing more complicated and are not necessary with typescript

File: src/components/context/ChatContext.tsx [hunk 13]
```diff
@@ -15,91 +14,74 @@ export const ChatContext = React.createContext<ChatContextType | undefined>(
...
+
+interface BaseChatMessage {
   from: string;
   text: string;
+  ts_utc: Time;
```
[the remark this answers is not in the document]
Comment (maxb94) [reply src/components/context/ChatContext.tsx:24]: privatechats > .. > chats > .. > has it as a firbase timestamp

File: src/components/context/ChatContext.tsx [hunk 14]
```diff
@@ -15,91 +14,74 @@ export const ChatContext = React.createContext<ChatContextType | undefined>(
...
+}
+
+type MessageBuilderInput = BaseMessageBuilderInput &
+  (
+    | { type: RestrictedMessageType | "private"; to: string }
+    | { type: "global" }
+  );
+
+// @debt typing this can be typed better
+function buildMessage(data: MessageBuilderInput): ChatMessage {
```
[the remark this answers is not in the document]
Comment (maxb94) [reply src/components/context/ChatContext.tsx:61]: Nope! any types should be avoided and `to: undefined` is 💩.. Also mutating message variable is unnecessary.

File: src/components/molecules/TablesUserList/TablesUserList.tsx [hunk 15]
```diff
@@ -73,20 +73,22 @@ const TablesUserList: React.FunctionComponent<PropsType> = ({
...
     }
-  }, [user, setSeatedAtTable, usersById, venueName]);
+  }, [profile, setSeatedAtTable, user, usersById, venueName]);
 
   if (!users) {
     return <>Loading...</>;
   }
 
-  let seatedAtTableName = "";
+  const seatedAtTableName = "";
```
[the remark this answers is not in the document]
Comment (maxb94) [reply src/components/molecules/TablesUserList/TablesUserList.tsx:89]: indeed. Need to have a eslint rule to not redeclare variables, seatedAtTableName is a prop

File: src/components/organisms/Chatbox/Chatbox.tsx [hunk 16]
```diff
@@ -60,28 +53,29 @@ const Chatbox: React.FunctionComponent<PropsType> = ({
...
     privateChats &&
-    (isInProfileModal ? privateChats : privateChats.concat(chats));
+    (isInProfileModal ? privateChats : [...privateChats, ...chats]);
 
   let chatsToDisplay =
     listOfChats &&
     listOfChats
       .filter(isChatValid)
-      .filter((chat: any) =>
+      .filter((chat: typeof listOfChats[number]) =>
```
[the remark this answers is not in the document]
Comment (maxb94) [reply src/components/organisms/Chatbox/Chatbox.tsx:62]: wrote this before adding the types i think. now removed.

File: src/components/organisms/Room/Room.tsx [hunk 17]
```diff
@@ -72,12 +81,10 @@ const Room = ({ roomName, setUserList, capacity = undefined }) => {
...
```
[the remark this answers is not in the document]
Comment (maxb94) [reply src/components/organisms/Room/Room.tsx:85]: the twillio types suggest no .stop method.

File: src/components/organisms/UserProfileModal/UserProfileModal.tsx [hunk 18]
```diff
@@ -29,17 +28,13 @@ const UserProfileModal: React.FunctionComponent<PropTypes> = ({
...
-    fullUserProfile = {
-      ...userProfile,
-      id: usersordered.find(
-        (u: User) => u.pictureUrl === userProfile.pictureUrl
-      )?.id,
-    };
-  } else {
-    fullUserProfile = userProfile;
-  }
+  // const fullUserProfile = !userProfile.id
```
[the remark this answers is not in the document]
Comment (maxb94) [reply src/components/organisms/UserProfileModal/UserProfileModal.tsx:31]: not sure. Left it in for posterity.

File: src/components/venues/Jazzbar/components/ReactionList.tsx (hunk 6 above)
Comment (maxb94) [reply src/components/venues/Jazzbar/components/ReactionList.tsx:59]: yes

File: src/hooks/useConnectCurrentEvent.ts (hunk 8 above)
Comment (maxb94) [reply src/hooks/useConnectCurrentEvent.ts:23]: that is what we want it to do. Can't put a if(!user)return due to hooks. Alternative is to provide an argument to the hook as (userId:string) and only call it in a component where user is defined, but this seems unnecessary.

File: src/types/Firestore.ts (hunk 10 above)
Comment (maxb94) [reply src/types/Firestore.ts:21]: indeed

File: src/types/PartyMapVenue.ts (hunk 11 above)
Comment (maxb94) [reply src/types/PartyMapVenue.ts:6]: yes

File: src/components/molecules/TablesUserList/TablesUserList.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (maxb94) [reply src/components/molecules/TablesUserList/TablesUserList.tsx]: fixed

File: src/components/context/ChatContext.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (mathieud) [reply src/components/context/ChatContext.tsx:61]: `undefined` was never allowed, the calls had to match one of the definitions above.

**Pull request merged**